tsl/src/continuous_aggs/invalidation.c

```diff
@@ -0,0 +1,512 @@
+	{
+		/*
+		 * Need to cut in right end:
+		 *
+		 *     |------|
+		 * [++++++]
+		 *
+		 * [++]
+		 */
+		tup1 = create_invalidation_tup(tupdesc,
+									   cagg_hyper_id,
+									   entry->modification_time,
+									   entry->lowest_modified_value,
+									   refresh_window->start - 1);
+		result = INVAL_CUT;
+	}
+	else if (entry->lowest_modified_value >= refresh_window->start &&
+			 entry->lowest_modified_value < refresh_window->end &&
+			 entry->greatest_modified_value >= refresh_window->end)
+	{
+		/*
+		 * Need to cut in left end:
+		 *
+		 * |------|
+		 *    [++++++++]
+		 *
+		 *         [+++]
+		 */
+		tup1 = create_invalidation_tup(tupdesc,
+									   cagg_hyper_id,
+									   entry->modification_time,
+									   refresh_window->end,
+									   entry->greatest_modified_value);
+		result = INVAL_CUT;
+	}
+
+	/* Insert any modifications into the cagg invalidation log */
+	if (tup1 || tup2)
+	{
+		CatalogSecurityContext sec_ctx;
+
+		ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+
+		if (tup1)
+		{
+			/* If the TID is set, we are updating an existing tuple, i.e., we
+			 * are processing and entry in the cagg log itself. Otherwise, we
+			 * are processing the hypertable invalidation log and need to
+			 * insert a new entry. */
+			if (tid)
+				ts_catalog_update_tid_only(state->cagg_log_rel, tid, tup1);
+			else
+				ts_catalog_insert_only(state->cagg_log_rel, tup1);
+
+			heap_freetuple(tup1);
+		}
+
+		if (tup2)
+		{
+			ts_catalog_insert_only(state->cagg_log_rel, tup2);
+			heap_freetuple(tup2);
+		}
+
+		ts_catalog_restore_user(&sec_ctx);
+	}
+
+	return result;
+}
+
+/*
+ * Process invalidations in the hypertable invalidation log.
+ *
+ * Copy and delete all entries from the hypertable invalidation log.  For the
+ * continuous aggregate that is getting refreshed, we also match the
+ * invalidation against the refresh window and perform additional processing
+ * (cutting or deleting); work that we'd otherwise have to do later in the
+ * cagg invalidation log.
+ *
+ * Note that each entry gets one copy per continuous aggregate in the cagg
+ * invalidation log (unless it matched the refresh window). These copied
+ * entries are later used to track invalidations across refreshes on a
+ * per-cagg basis.
+ *
+ * After this function has run, there are no entries left in the hypertable
+ * invalidation log.
+ */
+static void
+move_invalidations_from_hyper_to_cagg_log(const CaggInvalidationState *state,
+										  const InternalTimeRange *refresh_window)
+{
+	int32 hyper_id = state->cagg.data.raw_hypertable_id;
+	List *cagg_ids = get_cagg_ids(hyper_id);
+	int32 last_cagg_hyper_id = llast_int(cagg_ids);
+	ScanIterator iterator;
+	ListCell *lc;
+
+	Assert(list_length(cagg_ids) > 0);
+
+	/* Switch to a per-tuple memory context since we could be processing a lot
+	 * of invalidations (basically an unbounded amount). */
+	MemoryContextReset(state->per_tuple_mctx);
```

Comment:
Nit: Maybe I'm missing something, but it seems like this is resetting the memory context, but not switching to it. I see the switch below, but the comment is not entirely accurate.